tao_compiler/mlir/disc/utils/source_emitter.cc

```diff
@@ -0,0 +1,590 @@
+  } else if (type.isUnsignedInteger(64)) {
+    return "(unsigned long int)";
+  } else {
+    assert(false & "unsupported type for convert op.");
+  }
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::CosineOp>(Type type) {
+  if (type.isBF16() || type.isF16()) {
+    return "hcos";
+  } else if (type.isF32() || type.isF64()) {
+    return "cos";
+  } else {
+    assert(false & "unsupported type for cosine op.");
+  }
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::ExpOp>(Type type) {
+  if (type.isBF16() || type.isF16()) {
+    return "hexp";
+  } else if (type.isF32() || type.isF64()) {
+    return "exp";
+  } else {
+    assert(false & "unsupported type for exp op.");
+  }
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::FloorOp>(Type type) {
+  if (type.isBF16() || type.isF16()) {
+    return "hfloor";
+  } else if (type.isF32() || type.isF64()) {
+    return "floor";
+  } else {
+    assert(false & "unsupported type for floor op.");
+  }
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::IsFiniteOp>(Type type) {
+  if (type.isF32() || type.isF64()) {
+    return "isfinite";
+  } else {
+    assert(false & "unsupported type for isfinite op.");
+  }
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::LogOp>(Type type) {
+  if (type.isBF16() || type.isF16()) {
+    return "hlog";
+  } else if (type.isF32() || type.isF64()) {
+    return "log";
+  } else {
+    assert(false & "unsupported type for log op.");
+  }
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::Log1pOp>(Type type) {
+  if (type.isF32() || type.isF64()) {
+    return "log1p";
+  } else {
+    assert(false & "unsupported type for abs op.");
+  }
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::NegOp>(Type type) {
+  return "(-)";
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::NotOp>(Type type) {
+  return "(!)";
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::RsqrtOp>(Type type) {
+  if (type.isBF16() || type.isF16()) {
+    return "hrsqrt";
+  } else if (type.isF32() || type.isF64()) {
+    return "rsqrt";
+  } else {
+    assert(false & "unsupported type for rsqrt op.");
+  }
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::SqrtOp>(Type type) {
+  if (type.isBF16() || type.isF16()) {
+    return "hsqrt";
+  } else if (type.isF32() || type.isF64()) {
+    return "sqrt";
+  } else {
+    assert(false & "unsupported type for sqrt op.");
+  }
+}
+
+template <>
+std::string CUDAMathFuncName<lmhlo::TanhOp>(Type type) {
+  if (type.isF32() || type.isF64()) {
+    return "tanh";
+  } else {
+    assert(false & "unsupported type for tanh op.");
+  }
+}
+
+std::string MLIRType2CUDATypeStr(Type type) {
+  if (type.isBF16()) {
+    return "__nv_bfloat162";
+  } else if (type.isF16()) {
+    return "half";
+  } else if (type.isF32()) {
+    return "float";
+  } else if (type.isF64()) {
+    return "double";
+  } else if (type.isSignedInteger(1)) {
+    return "bool";
+  } else if (type.isSignedInteger(32)) {
```

Comment:
Int8 is not considered yet?

Reply:
Not implemented in current commit. More data types can be easily extended.

Reply:
> Done.

Thanks ~~ but it seems that the test case has not been add to `tao_build.py`, so not executed in PR test?